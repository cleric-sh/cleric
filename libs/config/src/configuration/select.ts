import {Primitive} from 'Misc/JSON/Primitive';
import 'reflect-metadata';
import {SENSITIVE_METADATA_KEY} from './decorators/Sensitive';
import {ConfigurationError} from './errors/ConfigurationError';

export type Selected<TConfig, TSelector extends Selector<TConfig>> = {
  [P in keyof TSelector]: P extends keyof TConfig ? NonNullable<TConfig[P]> : never;
};

export type Selector<TConfig extends Object> = {
  [P in keyof TConfig]?: TConfig[P] extends Primitive ? boolean : Selector<TConfig[P]>;
};

export function select<TConfig extends Object, TSelector extends Selector<TConfig>>(
  config: TConfig,
  selector: TSelector,
  log = false
): [Selected<TConfig, TSelector>, undefined | Selected<TConfig, TSelector>] {
  // eslint-disable-next-line
  const selected: any = {};
  // eslint-disable-next-line
  const loggable: any = {};

  for (const key in selector) {
    const value = config[key as keyof TConfig];
    const options = selector[key];

    if (typeof options === 'object') {
      const [selectedValue, loggableValue] = select(value, options as Selector<typeof value>);
      selected[key] = selectedValue;
      loggable[key] = loggableValue;
      continue;
    }

    // noinspection SuspiciousTypeOfGuard
    if (typeof options === 'boolean') {
      if (!value && options === true) {
        throw new ConfigurationError(
          `Configuration setting '${key}' was required but not provided by any source.`
        );
      }

      selected[key] = value;

      const isProtected = Reflect.hasMetadata(SENSITIVE_METADATA_KEY, config, key);
      loggable[key] = isProtected ? '[**********]' : value;
    }
  }

  return [selected, loggable];
}
