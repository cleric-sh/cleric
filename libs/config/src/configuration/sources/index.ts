import {Misc} from 'ts-toolbelt';

export interface ISource {
  get(configObject: Object, ...path: string[]): undefined | Misc.JSON.Value;
}

export type SourceKey = 'ConfigFile' | 'PlatformSecret' | 'ProcessEnv' | 'ServicesSecret';

export type Sources = Array<[SourceKey, ISource | undefined]>;

export * from './AwsSecretSource';
export * from './ConfigFileSource';
export * from './ProcessEnvSource';
