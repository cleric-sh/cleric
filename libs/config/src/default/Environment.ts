import {Deprecated} from '../configuration/decorators/Deprecated';
import {Use} from '../configuration/decorators/Use';

/**
 * Environment contains default that are required in order to load default for
 * other configuration classes.
 *
 * Values in Environment are loaded directly from ProcessEnv and ConfigFile.
 */
@Use('ProcessEnv', 'ConfigFile')
export class Environment {
  applicationName = '';
  region = 'eu-west-1';
  env = 'TEST';

  get isProduction() {
    return !!(this.env && this.env.toUpperCase() === 'PROD');
  }

  get platformKeysSecretName() {
    // Todo: this key should be provided via env var defined in task defs.
    return this.isProduction ? 'TibberPlatformKeysProduction' : 'TibberPlatformKeysDevelopment';
  }

  get servicesSecretName() {
    // Todo: this key should be provided via env var defined in task defs.
    return this.isProduction ? 'TibberServicesProduction' : 'TibberServicesDevelopment';
  }

  get environmentPrefix() {
    return this.isProduction ? 'prod' : 'test';
  }

  @Deprecated("use 'environmentPrefix' instead.")
  get awsResourcePrefix() {
    return this.environmentPrefix;
  }
}
