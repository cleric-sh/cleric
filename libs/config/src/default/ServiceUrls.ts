import {AwsSecretKey} from '../configuration/decorators/AwsSecretKey';
import {Use} from '../configuration/decorators/Use';

/**
 * Configuration object for service discovery.
 *
 * Values in ServiceUrls are loaded directly from ServicesSecret in AWS Secret Manager,
 * but can be overridden by sources in parent classes with a higher precedence.
 */
@Use('ServicesSecret')
export class ServiceUrls {
  @AwsSecretKey('tibber_app_gateway') appGateway = '';
  @AwsSecretKey('tibber_pyanalytics') analytics = '';
  @AwsSecretKey('tibber_bankid') bankId = '';
  @AwsSecretKey('tibber_cersei') cersei = '';
  @AwsSecretKey('tibber_customer') customer = '';
  @AwsSecretKey('tibber_pydemandresponse') demandResponse = '';
  @AwsSecretKey('tibber_easee') easee = '';
  @AwsSecretKey('tibber_edi_server_de') ediDe = '';
  @AwsSecretKey('tibber_edi_server_no') ediNo = '';
  @AwsSecretKey('tibber_edi_server') edi = '';
  @AwsSecretKey('tibber_elhub') elhub = '';
  @AwsSecretKey('tibber_email') email = '';
  @AwsSecretKey('tibber_energy_market') energyMarket = '';
  @AwsSecretKey('tibber_home_control') homeControl = '';
  @AwsSecretKey('tibber_invite') invite = '';
  @AwsSecretKey('tibber_iot') iot = '';
  @AwsSecretKey('tibber_pyml') ml = '';
  @AwsSecretKey('tibber_netarea') netArea = '';
  @AwsSecretKey('tibber_notifications') notification = '';
  @AwsSecretKey('tibber_onboarding') onboarding = '';
  @AwsSecretKey('tibber_payment') payment = '';
  @AwsSecretKey('tibber_payment_signup') paymentSignup = '';
  @AwsSecretKey('tibber_push_notifications') pushNotification = '';
  @AwsSecretKey('tibber_sms') sms = '';
  @AwsSecretKey('tibber_woocommerce_de') storeDe = '';
  @AwsSecretKey('tibber_woocommerce_no') storeNo = '';
  @AwsSecretKey('tibber_woocommerce_se') storeSe = '';
  @AwsSecretKey('tibber_subscription') subscription = '';
  @AwsSecretKey('tibber_telemetry') telemetry = '';
  @AwsSecretKey('tibber_timeseries') timeSeries = '';
  @AwsSecretKey('tibber_trading') trading = '';
  @AwsSecretKey('tibber_wallet') wallet = '';
}
