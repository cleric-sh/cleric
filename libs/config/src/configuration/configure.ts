import {Selector, select} from './select';

export interface ILogger {
  (message: string): void;
}

export interface IConfigArgs {
  logger?: ILogger;
}

export function configure<TArgs extends IConfigArgs, TConfig extends Object>(
  config: (args: TArgs) => TConfig,
  defaultArgs: TArgs
) {
  return function <TSelector extends Selector<TConfig>>(
    requires: TSelector,
    args: TArgs = defaultArgs
  ) {
    const configuration = config(args);
    const {logger} = args;

    const [selected, loggable] = select(configuration, requires, !!logger);

    if (logger) {
      const json = JSON.stringify(loggable, null, 2);
      logger(`SELECTED CONFIGURATION: ${json}`);
    }

    return selected;
  };
}
