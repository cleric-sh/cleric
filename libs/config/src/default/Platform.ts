import {AwsSecretKey} from '../configuration/decorators/AwsSecretKey';
import {Sensitive} from '../configuration/decorators/Sensitive';
import {Use} from '../configuration/decorators/Use';
import {Environment} from './Environment';

/**
 * Configuration object for platform default, such as DB connection strings.
 *
 * Values in Platform are loaded directly from PlatformSecret in AWS Secret Manager,
 * but can be overridden by sources in parent classes with a higher precedence.
 */
@Use('PlatformSecret')
export class Platform {
  constructor(private readonly core: Environment) {}

  @AwsSecretKey('connectionStringNodeJs')
  private _postgresConnection = '';

  @Sensitive()
  get postgresConnection() {
    const name = this.core.applicationName;
    return name ? this._postgresConnection + `?application_name=${name}` : this._postgresConnection;
  }

  @Sensitive()
  @AwsSecretKey('redisConfiguration')
  redisConnection = '';
}
