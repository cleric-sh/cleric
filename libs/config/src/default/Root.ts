import {Use} from '../configuration/decorators/Use';
import {Environment} from './Environment';
import {Platform} from './Platform';
import {ServiceUrls} from './ServiceUrls';

/**
 * Top level configuration object, containing all other nested configuration default.
 */
@Use('ProcessEnv', 'ConfigFile')
export class Root {
  constructor(readonly environment: Environment) {}

  get idpDiscoveryUrl() {
    // Todo: this key should be provided via AWS Secret Manager and moved to 'Platform'.
    return this.environment.isProduction
      ? 'https://thewall.tibber.com/.well-known/openid-configuration'
      : 'http://auth.tibberdev.com/.well-known/openid-configuration';
  }

  platform = new Platform(this.environment);
  serviceUrls = new ServiceUrls();
}
