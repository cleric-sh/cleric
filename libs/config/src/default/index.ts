import {configure as configureAws} from 'tibber-aws';
import {IConfigArgs} from '../configuration/configure';
import {populate} from '../configuration/populate/populate';
import {AwsSecretSource, Sources} from '../configuration/sources';
import {ConfigFileSource} from '../configuration/sources';
import {ProcessEnvSource} from '../configuration/sources';
import {Environment} from './Environment';
import {Root} from './Root';

export interface DefaultConfigArgs extends IConfigArgs {
  configFilePath: string;
}

export const DefaultConfigArgs: DefaultConfigArgs = {
  configFilePath: './config.json',
};

export function DefaultConfig(args: DefaultConfigArgs) {
  const {configFilePath} = args;

  /**
   * 'sources' determines the order of precedence of values. Values from
   * higher sources are preferred to lower sources.
   */
  const sources: Sources = [
    ['ProcessEnv', new ProcessEnvSource(['TIBBER'])],
    ['ConfigFile', new ConfigFileSource(configFilePath)],
  ];

  /**
   * Remarks: Ideally, we wouldn't need to populate Environment first, and we could just
   * populate a single configuration object, but since AwsSecretSources load different
   * secret keys based on whether or not 'env' is 'PROD' or not, we need to load them
   * first. In the future, we should consider running the production cluster in a separate
   * production AWS project, so that the Secret Keys can be static. Then the 'Environment'
   * configuration object can be rolled into Root and loaded in a single call to 'populate'..
   */
  const environment = populate(sources, Environment);

  const {platformKeysSecretName, region, servicesSecretName} = environment;

  configureAws({region});

  sources.push(['PlatformSecret', new AwsSecretSource(platformKeysSecretName)]);
  sources.push(['ServicesSecret', new AwsSecretSource(servicesSecretName)]);

  return populate(sources, Root, environment);
}

export default DefaultConfig;
