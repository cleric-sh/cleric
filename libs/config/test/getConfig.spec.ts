import {getSecretCollection} from 'tibber-aws';
import getConfig from '../src';
import {Selector} from '../src/configuration/select';
import {Root} from '../src/default/Root';
import {createJsonFile} from './_tools/createJsonFile';

/**
 * This test suite requires authenticated AWS credentials to be present, either as ENV VARS or in ~/.aws/credentials.
 */
describe('getConfig', () => {
  const expectedAppName = 'tibber-configuration-tests';

  process.env.TIBBER__APPLICATION_NAME = expectedAppName;
  process.env.TIBBER__SERVICE_URLS__ANALYTICS = 'http://url.from.process.env/analytics';

  createJsonFile('./dist/config.json', {
    platform: {
      redisConnection: 'redis-connection-from-config-file',
    },
    serviceUrls: {
      analytics: 'http://url.from.config.file/analytics',
      cersei: 'http://url.from.config.file/cersei',
    },
  });

  const selected: Selector<Root> = {
    environment: {
      applicationName: true,
      env: true,
      environmentPrefix: true,
    },
    platform: {
      postgresConnection: true,
      redisConnection: true,
    },
    serviceUrls: {
      analytics: true,
      appGateway: true,
      cersei: false,
    },
  };

  describe('test config', () => {
    let actualMessageWrittenToLog = '';

    const config = getConfig(selected, {
      configFilePath: './dist/config.json',
      logger: (message: string) => {
        actualMessageWrittenToLog = message;
      },
    });

    const platform = getSecretCollection('TibberPlatformKeysDevelopment') || {};
    const serviceUrls = getSecretCollection('TibberServicesDevelopment') || {};

    it('loads expected values from sources', () => {
      expect(config).toStrictEqual({
        environment: {
          applicationName: expectedAppName,
          env: 'TEST',
          environmentPrefix: 'test',
        },
        platform: {
          postgresConnection:
            platform['connectionStringNodeJs'] + `?application_name=${expectedAppName}`,
          redisConnection: 'redis-connection-from-config-file',
        },
        serviceUrls: {
          analytics: 'http://url.from.process.env/analytics',
          appGateway: serviceUrls['tibber_app_gateway'],
          cersei: 'http://url.from.config.file/cersei',
        },
      });
    });

    test('selected values are written to log', () => {
      expect(actualMessageWrittenToLog).toMatchSnapshot();
    });

    test('postgresConnection is redacted from log', () => {
      expect(actualMessageWrittenToLog).toContain('"postgresConnection": "[**********]"');
    });

    test('redisConnection is redacted from log', () => {
      expect(actualMessageWrittenToLog).toContain('"redisConnection": "[**********]"');
    });
  });

  describe('prod config', () => {
    process.env.TIBBER__ENV = 'PROD';

    let actualMessageWrittenToLog = '';

    const config = getConfig(selected, {
      configFilePath: './dist/config.json',
      logger: (message: string) => {
        actualMessageWrittenToLog = message;
      },
    });

    const platform = getSecretCollection('TibberPlatformKeysProduction') || {};
    const serviceUrls = getSecretCollection('TibberServicesProduction') || {};

    it('loads expected values from sources', () => {
      expect(config).toStrictEqual({
        environment: {
          applicationName: expectedAppName,
          env: 'PROD',
          environmentPrefix: 'prod',
        },
        platform: {
          postgresConnection:
            platform['connectionStringNodeJs'] + `?application_name=${expectedAppName}`,
          redisConnection: 'redis-connection-from-config-file',
        },
        serviceUrls: {
          analytics: 'http://url.from.process.env/analytics',
          appGateway: serviceUrls['tibber_app_gateway'],
          cersei: 'http://url.from.config.file/cersei',
        },
      });
    });

    test('selected values are written to log', () => {
      expect(actualMessageWrittenToLog).toMatchSnapshot();
    });

    test('postgresConnection is redacted from log', () => {
      expect(actualMessageWrittenToLog).toContain('"postgresConnection": "[**********]"');
    });

    test('redisConnection is redacted from log', () => {
      expect(actualMessageWrittenToLog).toContain('"redisConnection": "[**********]"');
    });
  });
});
