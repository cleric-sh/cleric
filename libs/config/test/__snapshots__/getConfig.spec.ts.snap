// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getConfig prod config selected values are written to log 1`] = `
"SELECTED CONFIGURATION: {
  \\"environment\\": {
    \\"applicationName\\": \\"tibber-configuration-tests\\",
    \\"env\\": \\"PROD\\",
    \\"environmentPrefix\\": \\"prod\\"
  },
  \\"platform\\": {
    \\"postgresConnection\\": \\"[**********]\\",
    \\"redisConnection\\": \\"[**********]\\"
  },
  \\"serviceUrls\\": {
    \\"analytics\\": \\"http://url.from.process.env/analytics\\",
    \\"appGateway\\": \\"http://tibber_app_gateway.production:3038\\",
    \\"cersei\\": \\"http://url.from.config.file/cersei\\"
  }
}"
`;

exports[`getConfig test config selected values are written to log 1`] = `
"SELECTED CONFIGURATION: {
  \\"environment\\": {
    \\"applicationName\\": \\"tibber-configuration-tests\\",
    \\"env\\": \\"TEST\\",
    \\"environmentPrefix\\": \\"test\\"
  },
  \\"platform\\": {
    \\"postgresConnection\\": \\"[**********]\\",
    \\"redisConnection\\": \\"[**********]\\"
  },
  \\"serviceUrls\\": {
    \\"analytics\\": \\"http://url.from.process.env/analytics\\",
    \\"appGateway\\": \\"http://tibber_app_gateway.dev:3038\\",
    \\"cersei\\": \\"http://url.from.config.file/cersei\\"
  }
}"
`;
